ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncServer.hs

```diff
@@ -20,63 +21,95 @@ import qualified Ouroboros.Storage.ChainDB.API as ChainDB
+      let ChainSyncServer server = chainSyncServerForReader tracer chainDB rdr
+      server
+
+-- | Chain Sync Server for blocks for a given a 'ChainDB'.
+--
+-- The local node-to-client protocol uses the chain sync mini-protocol with
+-- chains of full blocks (rather than a header \/ body split).
+--
+chainSyncBlocksServer
+    :: forall m blk. MonadSTM m
+    => Tracer m String
+    -> ChainDB m blk (Header blk)
+    -> ChainSyncServer blk (Point blk) m ()
+chainSyncBlocksServer tracer chainDB =
+    ChainSyncServer $ do
+      rdr <- ChainDB.newBlockReader chainDB
+      let ChainSyncServer server = chainSyncServerForReader tracer chainDB rdr
+      server
+
+
+-- | A chain sync server.
+--
+-- This is a version of
+-- 'Ouroboros.Network.Protocol.ChainSync.Examples.chainSyncServerExample' that
+-- uses a 'chainDB' and a 'Reader' instead of
+-- 'Ourboros.Network.ChainProducerState.ChainProducerState'.
+--
+-- All the hard work is done by the 'Reader's provided by the 'ChainDB'.
+--
+chainSyncServerForReader
+    :: forall m blk hdr b.
+       MonadSTM m
+    => HeaderHash blk ~ HeaderHash b -- b is used at either type blk or hdr
+    => Tracer m String
+    -> ChainDB m blk hdr
+    -> Reader  m b
+    -> ChainSyncServer b (Point blk) m ()
+chainSyncServerForReader _tracer chainDB rdr =
+    idle'
   where
-    idle :: Reader m (Header blk)
-         -> ServerStIdle (Header blk) (Point blk) m ()
-    idle rdr =
+    idle :: ServerStIdle b (Point blk) m ()
+    idle =
       ServerStIdle {
-        recvMsgRequestNext   = handleRequestNext rdr,
-        recvMsgFindIntersect = handleFindIntersect rdr,
+        recvMsgRequestNext   = handleRequestNext,
+        recvMsgFindIntersect = handleFindIntersect,
         recvMsgDoneClient    = return ()
       }
 
-    idle' :: Reader m (Header blk)
-          -> ChainSyncServer (Header blk) (Point blk) m ()
-    idle' = ChainSyncServer . return . idle
+    idle' :: ChainSyncServer b (Point blk) m ()
+    idle' = ChainSyncServer (return idle)
 
-    handleRequestNext :: Reader m (Header blk)
-                      -> m (Either (ServerStNext (Header blk) (Point blk) m ())
-                                (m (ServerStNext (Header blk) (Point blk) m ())))
-    handleRequestNext rdr = do
+    handleRequestNext :: m (Either (ServerStNext b (Point blk) m ())
+                                (m (ServerStNext b (Point blk) m ())))
+    handleRequestNext = do
       mupdate <- ChainDB.readerInstruction rdr
-      tip     <- atomically $ ChainDB.getTipPoint chainDB
+      tip     <- atomically $ castPoint <$> ChainDB.getTipPoint chainDB
       return $ case mupdate of
-        Just update -> Left  $ sendNext rdr tip update
-        Nothing     -> Right $ sendNext rdr tip <$> ChainDB.readerInstructionBlocking rdr
+        Just update -> Left  $ sendNext tip update
+        Nothing     -> Right $ sendNext tip <$>
+                                 ChainDB.readerInstructionBlocking rdr
```

Comment:
^ The `readerInstructionBlocking` will block for some time, I think it's possible that when it wakes the `tip` will also be different.

Reply:
Oh that's a good point, we should review that. That's independent of this change. File an issue.

Reply:
Done https://github.com/input-output-hk/ouroboros-network/issues/661